File: [BFS]BJ_2146.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>
#include <deque>

using namespace std;
int n;
int arr[100][100];
bool check[100][100];
int dy[4] = {-1, 0, 1, 0};
int dx[4] = {0, -1, 0, 1};

int bfs() {
    int cat = 1;
    int sol = 10000;
    for (int i=0;i<n;i++){
        for (int j=0;j<n;j++) {
            if (arr[i][j] != 1) continue; // 1이 아니라면 탐색 시작할 필요 없음
            cat += 1; // group categorizing index, 2번 부터 형성
            deque<pair<int, int>> q1; // BFS 그룹 형성을 위한 배열
            deque<tuple<int,int, int>> q2; // BFS 현재 그룹 queue
            q1.push_back(make_pair(i,j));
            q2.push_back(make_tuple(0,i,j));
            arr[i][j] = cat; // 그룹 원소 배열 형성
            while (q1.size()) {
                int y, x, ny, nx;
                y = q1.front().first;
                x = q1.front().second;
                q1.pop_front();
                for (int k=0;k<4;k++) {
                    ny = y+dy[k];
                    nx = x+dx[k];
                    if ((ny<0) || (ny>=n) || (nx<0) || (nx>=n)) continue;
                    if (arr[ny][nx] != 1) continue;
                    q1.push_back(make_pair(ny,nx));
                    q2.push_back(make_tuple(0,ny,nx));
                    arr[ny][nx] = cat;
                }
            }
            fill(&check[0][0], &check[99][100], false);
            bool flag = false;
            while (q2.size()) {
                int cost, y, x, ny, nx;
                cost = get<0>(q2.front());
                y = get<1>(q2.front());
                x = get<2>(q2.front());
                q2.pop_front();
                if (flag || (cost >= sol)) break;
                for (int k=0;k<4;k++) {
                    ny = y + dy[k];
                    nx = x + dx[k];
                    if ((ny<0) || (ny>=n) || (nx<0) || (nx>=n)) continue;
                    if (check[ny][nx]) continue;
                    if (arr[ny][nx] != 0) {
                        if (arr[ny][nx] != cat) {
                            sol = min(sol, cost);
                            flag = true;
                            break;
                        }
                        else continue;
                    }
                    q2.push_back(make_tuple(cost+1, ny, nx));
                    check[ny][nx] = true;
                }
            }
        }
    }
    return sol;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    // 입력값 처리
    cin >> n;
    for (int i=0;i<n;i++) {
        for (int j=0;j<n;j++) {
            cin >> arr[i][j];
        }
    }

    cout << bfs() << '\n';
    return 0;
}

File: [BFS]BJ_2146.py
import sys
from collections import deque

rdln = sys.stdin.readline
dy, dx = [-1, 0, 1, 0], [0, -1, 0, 1]

def input_data():
    n = int(rdln())
    arr = [list(map(int, rdln().split())) for _ in range(n)]
    return n, arr

def bfs():
    cat = 1
    sol = 10_000
    for i in range(n):
        for j in range(n):
            if arr[i][j] == 1: # arr[i][j] == 0이면 바다 이므로 skip
                cat += 1 # 섬번호 category
                q1 = deque([(i,j)]) # 현재 섬의 영역 확인
                q2 = deque([(0,i,j)]) # 다른 섬과의 거리를 확인 하기 위한 queue
                arr[i][j] = cat # 현재 섬번호 저장
                while q1: # 현재 섬 영역 확인 및 다른섬과의 거리 0으로 초기화
                    y, x = q1.popleft()
                    for k in range(4):
                        ny, nx = y + dy[k], x + dx[k]
                        if not (0<=ny<n and 0<=nx<n): continue
                        if arr[ny][nx] != 1: continue
                        q1.append((ny, nx))
                        q2.append((0, ny, nx))
                        arr[ny][nx] = cat
                
                check = [[0]*100 for _ in range(100)] # visit 배열
                flag = False
                while q2: # 현재 섬에서 다른 섬으로의 거리 파악 
                    cost, y, x = q2.popleft()
                    # check[y][x] = 1
                    if flag or (cost >= sol):
                        break
                    for k in range(4):
                        ny, nx = y + dy[k], x + dx[k]
                        if not (0<=ny<n and 0<=nx<n): continue
                        if check[ny][nx]: continue
                        if arr[ny][nx] != 0:
                            if arr[ny][nx] != cat: 
                                sol = min(sol, cost)
                                flag = True
                                break
                            else:
                                continue
                        q2.append((cost+1, ny, nx))
                        check[ny][nx] = 1
    return sol

n, arr = input_data()
sys.stdout.write(f"{bfs()}\n")
